### PR TITLE
Slave proto: introduce proxies for RemoteCommand, FileWriter and FileReader

### PR DESCRIPTION
**Pull request by tardyp**
These changes for http://trac.buildbot.net/ticket/2437 remove the pb.Referenceable in buildbot protocol.

Introducing proxies that are documented in the cls-protocols.rst file

Despite the size of the patch, there is so much new code actually, just a bunch of refactorization
A new integration test to make sure that transfers are actually working with the proxy.
Other integration tests where already taking care of RemoteCommand

### CODE REVIEW

**Comment by jaredgrubb** [comment]
Looks good to me!

**Comment by sa2ajj** [comment]
failing test:

```
[ERROR]
Traceback (most recent call last):
  File "/home/travis/build/buildbot/buildbot/master/buildbot/test/unit/test_buildslave_protocols_pb.py", line 188, in test_remoteStartCommand
    conn.remoteStartCommand(RCInstance, builder_name, commandID, remote_command, args)
  File "/home/travis/build/buildbot/buildbot/master/buildbot/buildslave/protocols/pb.py", line 192, in remoteStartCommand
    args = self.createArgsProxies(args)
  File "/home/travis/build/buildbot/buildbot/master/buildbot/buildslave/protocols/base.py", line 40, in createArgsProxies
    for k, v in args.iteritems():
exceptions.AttributeError: 'str' object has no attribute 'iteritems'

buildbot.test.unit.test_buildslave_protocols_pb.TestConnection.test_remoteStartCommand
```

**Comment by tardyp** [comment]
Added an example 'null' protocol for slaves leaving in the same reactor as the master.
I'm a bit disappointed that I only get 5% improvement of performance for the integration tests.

File: master/buildbot/process/buildstep.py [hunk 1]
```diff
@@ -496,9 +496,6 @@ def setRenderable(res, attr):
         except BuildStepFailed:
             self.results = FAILURE
 
-        except BuildSlaveTooOldError:
```
Comment (tardyp) [inline master/buildbot/process/buildstep.py]: I spent a faire amount of time debugging why my step went in exception with no message. the generic except Exception catches and reports the error much more nicely

**Comment by tardyp** [comment]
@djmitche @tomprince , this finishes the slave-proto gsoc (obvious appart from the amp proto). I'd appreciate some review to make sure this fits you vision

File: master/docs/developer/cls-protocols.rst [hunk 2]
```diff
@@ -16,10 +19,21 @@ to know about protocol calls or handle protocol specific exceptions.
...
+    .. py:attribute:: proxies
+
+        Dictionary containing mapping between ``Impl`` classes and ``Proxy`` class for this protocol
+        This may be overridden by subclass to declare its proxy implementations
+
+    .. py:method:: createArgsProxies(args)
+
+        :returns: shallow copy of args dictionary with proxies instead of impls
+
+        Helper methods that will use :attr:`proxies`, and replace ``Impl`` objects by specific ``Proxy`` counterpart.
```
Comment (djmitche) [inline master/docs/developer/cls-protocols.rst]: `method` (singular)

**Comment by djmitche** [comment]
This is really cool, and nicely implemented.

My *one* concern is that this continues the model of the protocol holding references to master-side objects (RemoteCommand, FileWriter, and FileReader).  Where in the past we inherited from `pb.Referencable`, we now inherit from `*Impl`.  So we get the level of abstraction we need to support non-PB protocols.  But are we then requiring protocol implementations to do some of the crazy reference-holding magic that PB itself does (where a master-side reference is kept until the reference to it on the slave goes away)?

I think the answer is "no", from what I see here, but figured I'd ask the question anyway :)

**Comment by djmitche** [comment]
Also, are proxies useful for anything but PB?  Should those move to the PB protocol?

**Comment by tardyp** [comment]
Le Wed Feb 04 2015 at 5:17:12 AM, Dustin J. Mitchell <
notifications@github.com> a écrit :

> This is really cool, and nicely implemented.
>
> My *one* concern is that this continues the model of the protocol holding
> references to master-side objects (RemoteCommand, FileWriter, and
> FileReader). Where in the past we inherited from pb.Referencable, we now
> inherit from *Impl. So we get the level of abstraction we need to support
> non-PB protocols.
>
I also had this concern. But I then had a bigger concern which is how we
support old slaves?
How do we implement the new buildslave package to be compatible with old
protocol, and a new protocol?

> But are we then requiring protocol implementations to do some of the crazy
> reference-holding magic that PB itself does (where a master-side reference
> is kept until the reference to it on the slave goes away)?
>
I actually think pb's referenceable concept is a good way of resolving the
problem of command id.
I did not realize the problem of the gc, but I think that can be solved.
If we don't do so, we'll anyway have to put put the problem of reference id in
the logic code.

Actually, we already have a command_id, for interrupt command, so it is
feasible.

Any way I think the problem is: do we keep legacy protocol support?

**Comment by djmitche** [comment]
I think we do keep legacy protocol support -- from what I can tell, your patch does so!  Did I miss something there?

My concerns are really about whether this is a good design looking forward.  My last comment mused about that, but I don't actually see any problem.  The process code supplies some *Impl objects to the protocol implementation, and it's up to that protocol implementation to manage their lifetime appropriately.  For protocols tied to a TCP connection, that's pretty easy.  For more "ephemeral" uses, the protocol may have some work to do to be sure it doesn't keep references to those objects for too long -- but the protocol is probably the right place for that work.

So I think the only remaining concerns I have are:

 * Are proxies useful for anything but PB? Should those move to the PB protocol?
 * Does this break legacy protocol support in a way that I'm not seeing?

**Comment by tardyp** [comment]
Yes my patch keeps the legacy protocol. The thing is I don't see how we can keep the legacy protocol, and not "model the protocol to hold references to master-side objects" 

Anyway, I think we have an agreement. I agree that the Proxy class are actually uneeded. I dont think I'll need them either for my wamp PoC.

So. rebased, and cleaned that part (https://github.com/buildbot/buildbot/commit/60a5917ee2354a55c7e5b932d363c4a59e44d241 is new)

File: master/buildbot/buildslave/protocols/base.py [hunk 3]
```diff
@@ -66,3 +76,39 @@ def remoteStartBuild(self, builderName):
 
     def remoteInterruptCommand(self, commandId, why):
         raise NotImplementedError
+
+
+# RemoteCommand base implementation and base proxy
```
Comment (tomprince) [inline master/buildbot/buildslave/protocols/base.py]: subclasses :frowning: http://www.youtube.com/watch?v=3MNVP9-hglc

Comment (tardyp) [inline master/buildbot/buildslave/protocols/base.py]: Well, this file was already based on subclassing.. and you mentored this gsoc project, afaik :)

If we'd followed all google's best practices, we'd rewrite buildbot in golang. :->

**Comment by sa2ajj** [comment]
Let's take it from here...

**Pull request merged**